safehttp/task.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"net/http"
+)
+
+type Task struct {
+	rw  http.ResponseWriter
+	req *IncomingRequest
+
+	cfg HandlerConfig
+
+	code   StatusCode
+	header Header
+
+	written      bool
+	writtenError bool
+}
+
+// NewTask creates a ResponseWriter from a safehttp.Dispatcher, an
+// http.ResponseWriter and a reference to the current IncomingRequest being served.
+// The IncomingRequest will only be used by the commit phase, which only runs when
+// the ServeMux is used, and can be passed as nil in tests.
+// TODO: remove the IncomingRequest parameter.
+
+// ProperNewTask TODO
+func BrokenNewTask(rw http.ResponseWriter, dispatcher Dispatcher) *Task {
+	return &Task{
+		cfg:    HandlerConfig{Dispatcher: dispatcher},
+		rw:     rw,
+		header: newHeader(rw.Header()),
+	}
+}
+
+type HandlerConfig struct {
+	Handler      Handler
+	Dispatcher   Dispatcher
+	Interceptors []ConfiguredInterceptor
+}
+
+func ProcessRequest(cfg HandlerConfig, rw http.ResponseWriter, req *http.Request) {
```

Comment:
Does this need to be exported?

Reply:
Not in this form most likely. We'd rather have a "safe handler factory" in the future.